### PR TITLE
Add the repoupdater tool to this repository

### PR DESCRIPTION
**Pull request by bloodearnest**
The only change from the original is to use the same auth method, and to
exclude the explit repos as opposed to just regex for '-research'

### CODE REVIEW

File: repoupdater.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    for repo in repos:
+        path = os.path.join(BASE_PATH, repo.name)
+        print("-" * 80)
+        print(repo.name)
+        os.chdir(path)
+        subprocess.run(["git", "checkout", branch], check=True)
+        subprocess.run(["git", "push", "-u", "origin", branch], check=True)
+        try:
+            repo.create_pull(head=branch, base="master", title=title, body="")
```
Comment (madwort) [inline repoupdater.py:96]: I realise this is a can of worms, but is `base` guaranteed to be `master`, or could it be `main`?

File: repoupdater.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        update()
+    elif args.subcommand == "exec":
+        exec_([args.command] + args.args)
+    elif args.subcommand == "pull-request":
+        pull_request(args.branch, args.title)
+    else:
+        assert False, args.subcommand
+
+
+def list_():
```
Comment (madwort) [inline repoupdater.py:51]: how about `list_repos()` instead of `list_()`?

File: repoupdater.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+        if os.path.exists(path):
+            os.chdir(path)
+            subprocess.run(["git", "checkout", "master"], check=True)
+            subprocess.run(["git", "pull"], check=True)
+        else:
+            subprocess.run(["git", "clone", repo.ssh_url, path], check=True)
+
+
+def exec_(argv):
```
Comment (madwort) [inline repoupdater.py:76]: how about something like `exec_all_repos()`?

Comment (madwort) [reply repoupdater.py:76]: or just `exec_repos()`, as all these functions run against all repos

File: repoupdater.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+import subprocess
+
+from github import Github
+from github.GithubException import GithubException
+import yaml
+
+import client
+
+ORG_NAME = "OpenSAFELY"
+REPO_NAME_PATTERN = "^.*-research$"
```
Comment (madwort) [inline repoupdater.py:15]: unused var?

File: repoupdater.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        if os.path.exists(path):
+            os.chdir(path)
+            subprocess.run(["git", "checkout", "master"], check=True)
+            subprocess.run(["git", "pull"], check=True)
+        else:
+            subprocess.run(["git", "clone", repo.ssh_url, path], check=True)
+
+
+def exec_(argv):
+    for path in sorted(glob.glob(os.path.join(BASE_PATH, "*"))):
```
Comment (madwort) [inline repoupdater.py:77]: ah, ok, it's not all repos, it's all subdirectories, which is a bit counterintuitive to me. Would it be better to do the equivalent of https://github.com/opensafely/sysadmin/pull/7/files#diff-fdb7becf8fb97e0422febce02a3042d9a1d3db6539ed13d30f53f971c9c65989R63-R64 here?

File: repoupdater.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+def main():
+    parser = argparse.ArgumentParser(prog="repoupdater")
+    subparsers = parser.add_subparsers(help="sub-command help", dest="subcommand")
+    list_parser = subparsers.add_parser("list", help="list all repos")
+    update_parser = subparsers.add_parser(
+        "update", help="update all repos via pull or clone"
+    )
+    exec_parser = subparsers.add_parser(
+        "exec", help="execute command against all repos"
```
Comment (madwort) [inline repoupdater.py:27]: see other comment (currently all subdirs, not all repos)

File: README.repoupdater.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+$ python repoupdater.py exec -- sed -i s/suppresion/suppression/ codelists/codelists.txt
+
+# Update study definitions (need to use sed, for wildcard support)
+$ sed -i s/suppresion/suppression/ research/*/analysis/*.py
+
+# Add, commit, push
+# Note that if nothing has been added, the commit will error -- ignore this
+$ python repoupdater.py exec git add .
+$ python repoupdater.py exec -- git commit -m "Fix typo in codelist name"
+$ python repoupdater.py exec -- git push -u origin fix-suppression-codelists
```
Comment (madwort) [inline README.repoupdater.md:35]: nb. the `pull_request` action also does `push -u origin` so it seemsa bit unnecessary to have this here, but maybe it's fine. 
(follow-up thought - is it worth having a `repoupdater.py commit` function that does `commit -m && push -u` ?)

File: README.repoupdater.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+# Create PRs
+# Note that errors will be reported for repos where nothing has changed
+$ python repoupdater.py pull-request fix-suppression-codelists "Fix typo in codelist name"
```
Comment (madwort) [inline README.repoupdater.md:39]: will this title typically match the title for `commit -m`? can we fold this into one big helper that does the full `commit -m && push -u && create_pull `?

File: repoupdater.py (hunk 1 above)
Comment (bloodearnest) [reply repoupdater.py:96]: ugh, yeah, probably need to handle that

File: repoupdater.py (hunk 4 above)
Comment (bloodearnest) [reply repoupdater.py:15]: will delete

File: repoupdater.py (hunk 2 above)
Comment (bloodearnest) [reply repoupdater.py:51]: yeah, I just copied this from opensafely/repoupdater and made it use the explict config, so these are the original names. But easy enough to fix.

File: repoupdater.py (hunk 6 above)
Comment (bloodearnest) [reply repoupdater.py:27]: subdirs ~= all checked out repos?

File: README.repoupdater.md (hunk 7 above)
Comment (bloodearnest) [reply README.repoupdater.md:35]: yeah, that's been discussed, but separate piece of work I think.

File: README.repoupdater.md (hunk 8 above)
Comment (bloodearnest) [reply README.repoupdater.md:39]: Yep, as a later addition - I think. Main goal of this PR was to bring the existing repoupdater code into sysadmin and use the explicit config.

File: repoupdater.py (hunk 2 above)
Comment (madwort) [reply repoupdater.py:51]: ah right, do it in a follow-up PR then!

File: repoupdater.py (hunk 3 above)
Comment (madwort) [reply repoupdater.py:76]: do it a follow-up PR

**Review by madwort — APPROVED** [review]
lots of suggestions for improvements, but as it's a transfer from another repo rather than new code let's get the transfer done & then think about making changes.

File: repoupdater.py (hunk 6 above)
Comment (madwort) [reply repoupdater.py:27]: yeah, the set of subdirs should contain the set of repos, but also contain some other stragglers, e.g. if you ran `$ python repoupdater.py exec git add .` but you have a `./venv/` lying around? not a big deal, though, will fail in an obvious way.

**Pull request merged**